drivers/sensor/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (c) 2021 G-Technologies Sdn. Bhd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.winsen-sensor.com/sensors/co2-sensor/mh-z19b.html
+ */
+
+#define DT_DRV_COMPAT winsen_mhz19b
+
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <drivers/sensor.h>
+
+#include <drivers/sensor/mhz19b.h>
+#include "mhz19b.h"
+
+LOG_MODULE_REGISTER(mhz19b, CONFIG_SENSOR_LOG_LEVEL);
+
+static void mhz19b_uart_flush(const struct device *uart_dev)
+{
+	uint8_t c;
+
+	while (uart_fifo_read(uart_dev, &c, 1) > 0) {
+		continue;
+	}
+}
+
+static uint8_t mhz19b_checksum(const uint8_t *data)
+{
+	uint8_t cs = 0;
+
+	for (uint8_t i = 1; i < MHZ19B_BUF_LEN - 1; i++) {
+		cs += data[i];
+	}
+
+	return 0xff - cs + 1;
+}
+
+static int mhz19b_send_cmd(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx,
+			   const bool has_rsp)
+{
+	struct mhz19b_data *data = dev->data;
+	const struct mhz19b_cfg *cfg = dev->config;
+	int ret;
+
+	/* Make sure last command has been transferred */
+	ret = k_sem_take(&data->tx_sem, MHZ19B_WAIT);
+	if (ret) {
+		return ret;
+	}
+
+	data->cmd_idx = cmd_idx;
+	data->has_rsp = has_rsp;
+	k_sem_reset(&data->rx_sem);
+
+	uart_irq_tx_enable(cfg->uart_dev);
+
+	if (has_rsp) {
+		uart_irq_rx_enable(cfg->uart_dev);
+		ret = k_sem_take(&data->rx_sem, MHZ19B_WAIT);
+	}
+
+	return ret;
+}
+
+static inline int mhz19b_send_config(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx)
+{
+	struct mhz19b_data *data = dev->data;
+	int ret;
+
+	ret = mhz19b_send_cmd(dev, cmd_idx, true);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (data->rd_data[MHZ19B_RX_CMD_IDX] != mhz19b_cmds[data->cmd_idx][MHZ19B_TX_CMD_IDX]) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static inline int mhz19b_poll_data(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx)
+{
+	struct mhz19b_data *data = dev->data;
+	uint8_t checksum;
+	int ret;
+
+	ret = mhz19b_send_cmd(dev, cmd_idx, true);
+	if (ret < 0) {
+		return ret;
+	}
+
+	checksum = mhz19b_checksum(data->rd_data);
+	if (checksum != data->rd_data[MHZ19B_CHECKSUM_IDX]) {
+		data->data_valid = false;
+		return -EBADMSG;
+	}
+
+	data->data_valid = true;
+
+	switch (cmd_idx) {
+	case MHZ19B_CMD_IDX_GET_CO2:
+		data->data = sys_get_be16(&data->rd_data[2]);
+		break;
+	case MHZ19B_CMD_IDX_GET_RANGE:
+		data->data = sys_get_be16(&data->rd_data[4]);
+		break;
+	case MHZ19B_CMD_IDX_GET_ABC:
+		data->data = (uint16_t)data->rd_data[7];
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int mhz19b_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct mhz19b_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_CO2) {
+		return -ENOTSUP;
+	}
+
+	if (!data->data_valid) {
+		LOG_DBG("Checksum mismatch");
+		return -EINVAL;
+	}
```

Comment:
this is redundant: fetch already fails under this condition, so get should not be called.